ejb/src/main/java/org/wildfly/httpclient/ejb/RequestBuilder.java

```diff
@@ -0,0 +1,220 @@
+            headers.add(ACCEPT, INVOCATION_ACCEPT + "," + EJB_EXCEPTION);
+            headers.put(CONTENT_TYPE, INVOCATION.toString());
+            if (invocationId != null) {
+                headers.put(INVOCATION_ID, invocationId);
+            }
+            if (compressRequest) {
+                headers.put(CONTENT_ENCODING, GZIP.toString());
+            }
+            if (compressResponse) {
+                headers.put(ACCEPT_ENCODING, GZIP.toString());
+            }
+            headers.put(TRANSFER_ENCODING, CHUNKED.toString());
+        } else if (requestType == CREATE_SESSION) {
+            headers.add(ACCEPT, EJB_EXCEPTION.toString());
+            headers.put(CONTENT_TYPE, SESSION_OPEN.toString());
+        } else if (requestType == DISCOVER) {
+            headers.add(ACCEPT, EJB_DISCOVERY_RESPONSE + "," + EJB_EXCEPTION);
+        } else if (requestType != CANCEL_INVOCATION) {
+            throw new IllegalStateException();
+        }
+    }
+
+    private String getCreateSessionEjbRequestPath(final String prefix) {
+        final StringBuilder sb = new StringBuilder();
+        appendOperationPath(sb, prefix);
+        appendBeanPath(sb);
+        return sb.toString();
+    }
+
+    private String getDiscoverEjbRequestPath(final String prefix) {
+        final StringBuilder sb = new StringBuilder();
+        appendOperationPath(sb, prefix);
+        return sb.toString();
+    }
+
+    private String getCancelEjbInvocationRequestPath(final String prefix) {
+        final StringBuilder sb = new StringBuilder();
+        appendOperationPath(sb, prefix);
+        appendBeanPath(sb);
+        appendPath(sb, invocationId, false);
+        appendPath(sb, "" + cancelIfRunning, false);
+        return sb.toString();
+    }
+
+    private String getStartEjbInvocationRequestPath(final String prefix) {
+        final StringBuilder sb = new StringBuilder();
+        appendOperationPath(sb, prefix);
+        appendBeanPath(sb);
+        appendPath(sb, beanId, false);
+        appendPath(sb, view, false);
+        appendPath(sb, method.getName(), false);
+        for (final Class<?> param : method.getParameterTypes()) {
+            appendPath(sb, param.getName(), true);
+        }
+        return sb.toString();
+    }
+
+    private void appendBeanPath(final StringBuilder sb) {
+        appendPath(sb, locator.getAppName(), true);
+        appendPath(sb, locator.getModuleName(), true);
+        appendPath(sb, locator.getDistinctName(), true);
+        appendPath(sb, locator.getBeanName(), true);
+    }
+
+    private void appendOperationPath(final StringBuilder sb, final String prefix) {
+        if (prefix != null) {
+            sb.append(prefix);
+        }
+        appendPath(sb, "ejb", false);
+        appendPath(sb, "v" + version, false);
+        appendPath(sb, requestType.getPath(), false);
+    }
+
+    private static void appendPath(final StringBuilder sb, final String path, final boolean encode) {
```

Comment:
Some path parameters used are already prefixed with a forward slash (e.g. START_INVOCATION = "/invoke") so this will result in double slashes in the request context path?

Reply:
No, this is addressed later Richard. See my WIP branch.